lib/litani_report.py

```diff
@@ -297,6 +298,148 @@ def render(self, job_filter, template_name):
+                            temporary_report_dir / pipe["url"], env, job,
+                            gnuplot)
+
+                    tags = job["wrapper_arguments"]["tags"]
+                    description = job["wrapper_arguments"]["description"]
+                    if tags and "front-page-text" in tags:
+                        if "stdout" in job and job["stdout"]:
+                            front_page_outputs[description] = job
+
+            pipe_page = pipe_templ.render(run=run, pipe=pipe)
+            with litani.atomic_write(
+                    temporary_report_dir / pipe["url"] / "index.html") as handle:
+                print(pipe_page, file=handle)
+
+        dash_templ = env.get_template("dashboard.jinja.html")
+        page = dash_templ.render(
+            run=run, svgs=svgs,
+            litani_version=litani.VERSION,
+            litani_report_archive_path=litani_report_archive_path,
+            summary=get_summary(run), front_page_outputs=front_page_outputs)
+
+        with litani.atomic_write(temporary_report_dir / "index.html") as handle:
+            print(page, file=handle)
+
+        # The sequence is:
+        #
+        # Initially:
+        # |\_ report_data_dir
+        # |    |\_ 1234-abcd [report data that has just been written]
+        # |    \__ 5678-olde
+        # \__ html  -symlink->  report_data_dir/5678-olde
+        #
+        # 1. 1234-abcd gets locked
+        #
+        # 2. report_data_dir
+        #    |- 1234-abcd
+        #    |- 5678-olde
+        #    html  -symlink->  report_data_dir/5678-olde
+        #    html.987-xzy  -symlink->  1234-abcd
+        #
+        # 3. report_data_dir
+        #    |- 1234-abcd
+        #    |- 5678-olde
+        #    html  -symlink->  1234-abcd
+        #
+        # 4. 1234-abcd gets unlocked
+        #
+        # 5. 5678-olde gets marked as expired.
+
+        locked_dir = litani.LockableDirectory(temporary_report_dir)
+
+        temp_symlink_dir = self.report_dir.with_name(
+            self.report_dir.name + str(uuid.uuid4()))
+        os.symlink(temporary_report_dir, temp_symlink_dir)
+        os.rename(temp_symlink_dir, self.report_dir)
+
+        locked_dir.release()
+
+        try:
+            old_report_dir.expire()
+        except FileNotFoundError:
+            pass
+
+        litani.unlink_expired()
+
+
+
+async def release_html_dir(args):
+    lockable_dir = lib.litani.LockableDirectory(args.dir)
+    lockable_dir.release()
+
+
+async def acquire_html_dir(args):
+    remaining = args.timeout
+    while True:
+        html_dir = lib.litani.get_report_dir().resolve()
+        lockable_dir = lib.litani.LockableDirectory(html_dir)
+        try:
+            if lockable_dir.acquire():
+                print(html_dir)
+                sys.exit(0)
+        except FileNotFoundError:
+            pass
+        remaining -= 1
+        if remaining == 0:
+            sys.exit(1)
+        await asyncio.sleep(1)
```

Comment:
Why is the block of 1 second via `await asyncio.sleep(1)` being added?

Reply:
We want to retry every second, there's no need to retry as fast as possible?

`time.sleep` isn't the right call here because we're inside an `async` method. What `asyncio.sleep` does is to put the current green thread to sleep, but allow other green threads to proceed if they have work to do. Now in this particular case, there are actually no other green threads (the only litani command that is actually multi-threaded is `run-build`, since it both needs to run the command and do a bunch of monitoring like parallelism and so on). But it still looks very suspicious to have `time.sleep` in an async method, so I'd rather not have it